core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -239,6 +274,11 @@ public void run() {
         // the aggregator receives actual uses.
         rescheduleIdleTimer();
       }
+      if (graceLoadBalancer != null) {
+        // Exit grace period; timer already rescheduled above.
+        loadBalancer = graceLoadBalancer;
```

Comment:
Setting them needs to be done under the lock, to avoid racing with shutdown(). It would be possible to use volatile writes instead, but it would probably require a new "state" object that is created for each state transition. It'd be a pain, and I'm not sure how much we would gain from optimizing that path. Although it is true that it could probably reduce our tail latency.